### PR TITLE
fix: code coverage incomplete on redirect

### PR DESCRIPTION
**Pull request by fvclaus**
We have a frontend that, upon loading, redirects the user to Keycloak. The user logs in and is returned to the application. The code coverage written to the disk was incomplete and after hours of debugging I found the issue. The problem is the caching of references to the  `window.__coverage__` object. When the code that redirects the user is instrumented, the covered statements will be different for the initial redirect and the subsequent return so this plugin will correctly recognize them as different and will keep a reference to both coverage objects. When the code that redirects the user is not instrumented (because it is a library), the coverage object will be identical. Both coverage objects will have identical statement coverage and the current version of the code will throw away the second coverage object, keeping only the reference to the first one which is not updated anymore. This will result in a code coverage where only the statments until the call to the library are covered.

Before:
```
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |      50 |      100 |       0 |      50 |                   
 utils.js |      50 |      100 |       0 |      50 | 2                 
----------|---------|----------|---------|---------|-------------------
ERROR: Coverage for lines (50%) does not meet global threshold (100%)
```
After:

```
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |     100 |      100 |     100 |     100 |                   
 utils.js |     100 |      100 |     100 |     100 |                   
----------|---------|----------|---------|---------|-------------------
```

Keywords for Google: Code coverage different in `afterEach`.

### CODE REVIEW

File: .nycrc.json [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {
   "exclude": [
     "support-utils.js",
-    "task-utils.js"
+    "task-utils.js",
+    "support.js"
```
Comment (mschile) [inline .nycrc.json:5]: Why is this being excluded?

File: support.js [hunk 2]
```diff
@@ -72,13 +72,15 @@ const registerHooks = () => {
...
-      ) {
+      const existingCoverage = Cypress._.find(windowCoverageObjects, {
+        coverage: applicationSourceCoverage
+      })
+      if (existingCoverage) {
         // this application code coverage object is already known
         // which can happen when combining `window:load` and `before` callbacks
+        // it can also happen when the user leaves and returns to the application under test
+        // in which case we need to use new applicationSourceCoverage, because the old will not be updated anymore.
+        existingCoverage.coverage = applicationSourceCoverage
```
Comment (mschile) [inline support.js:83]: Does this cause any of the prior code coverage to be lost or does `applicationSourceCoverage` also contain the prior coverage?

Comment (fvclaus) [reply support.js:83]: It shouldn't. There is an equality comparison a few lines above, so only coverage objects with equivalent content should be replaced.

Comment (fvclaus) [reply support.js:83]: The coverage objects being replaced is not updated anymore and the new one replacing it has the some content.

File: .nycrc.json (hunk 1 above)
Comment (fvclaus) [reply .nycrc.json:5]: If you run the unit-tests, support.js is reported as having too little coverage, but support.js is not even being tested (only support-utils.js). I thought that this is an issue with my code, but I get the same result on the main branch.

**Comment by fvclaus** [comment]
@mschile , @cacieprins This MR has been hanging around for almost 4 months. Is there something I can do to get this merged?

**Comment by mschile** [comment]
Hi @fvclaus 👋, sorry for the delay. I will try to get back to this very soon.

**Review by mschile — APPROVED** [review]

**Pull request merged**